File: Dev/Tools/node-legacy/render_runtime_exports.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const { execFileSync } = require("child_process");

const ROOT = process.cwd();
const SOURCE = "Dev/Source/YAML/json/runtime_exports.yaml";

function loadYaml(relPath) {
    const script = [
        "import json, sys, yaml",
        "with open(sys.argv[1], encoding='utf-8') as handle:",
        "    data = yaml.safe_load(handle) or {}",
        "print(json.dumps(data, ensure_ascii=False))"
    ].join("\n");
    const stdout = execFileSync("python3", ["-c", script, path.join(ROOT, relPath)], {
        encoding: "utf8",
        maxBuffer: 4 * 1024 * 1024
    });
    return JSON.parse(stdout);
}

function normalizeSpecs(entries) {
    return Array.isArray(entries)
        ? entries.map(entry => ({
            key: String(entry?.key ?? "").trim(),
            path: String(entry?.path ?? "").replace(/\\/g, "/").trim()
        })).filter(entry => entry.key && entry.path)
        : [];
}

function runtimeModules(source) {
    const groups = {};
    for (const [groupId, entries] of Object.entries(source.runtime_modules ?? {})) {
        groups[groupId] = normalizeSpecs(entries);
    }
    return groups;
}

function commonjsRuntime(source) {
    const runtime = source.commonjs_runtime ?? {};
    return {
        entrypoints: normalizeSpecs(runtime.entrypoints),
        local_dependencies: normalizeSpecs(runtime.local_dependencies),
        data_dependencies: normalizeSpecs(runtime.data_dependencies)
    };
}

function main() {
    const source = loadYaml(SOURCE);
    const target = String(source.runtime_data ?? "z.automazioni/data/runtime/runtime_exports.json").trim();
    const payload = {
        generated_by: "render_runtime_exports",
        generated_from: SOURCE,
        source: SOURCE,
        purpose: "Registry runtime dei moduli caricati dal bridge DataviewJS e delle dipendenze CommonJS Templater.",
        commonjs_runtime: commonjsRuntime(source),
        runtime_modules: runtimeModules(source)
    };

    const outPath = path.join(ROOT, target);
    fs.mkdirSync(path.dirname(outPath), { recursive: true });
    fs.writeFileSync(outPath, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
    console.log(`Runtime exports data renderizzato: ${target}`);
}

main();

File: Dev/Tools/python/render_runtime_exports.py
#!/usr/bin/env python3

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

import yaml

sys.dont_write_bytecode = True

ROOT = Path(__file__).resolve().parents[3]
SOURCE = "Dev/Source/YAML/json/runtime_exports.yaml"


def normalize_specs(entries: Any) -> list[dict[str, str]]:
    if not isinstance(entries, list):
        return []
    specs: list[dict[str, str]] = []
    for entry in entries:
        if not isinstance(entry, dict):
            continue
        key = str(entry.get("key") or "").strip()
        path = str(entry.get("path") or "").replace("\\", "/").strip()
        if key and path:
            specs.append({"key": key, "path": path})
    return specs


def runtime_modules(source: dict[str, Any]) -> dict[str, list[dict[str, str]]]:
    groups: dict[str, list[dict[str, str]]] = {}
    modules = source.get("runtime_modules") if isinstance(source.get("runtime_modules"), dict) else {}
    for group_id, entries in modules.items():
        groups[str(group_id)] = normalize_specs(entries)
    return groups


def commonjs_runtime(source: dict[str, Any]) -> dict[str, list[dict[str, str]]]:
    runtime = source.get("commonjs_runtime") if isinstance(source.get("commonjs_runtime"), dict) else {}
    return {
        "entrypoints": normalize_specs(runtime.get("entrypoints")),
        "local_dependencies": normalize_specs(runtime.get("local_dependencies")),
        "data_dependencies": normalize_specs(runtime.get("data_dependencies")),
    }


def main() -> int:
    source_path = ROOT / SOURCE
    source = yaml.safe_load(source_path.read_text(encoding="utf-8")) or {}
    if not isinstance(source, dict):
        raise SystemExit(f"{SOURCE}: root YAML non valida")

    target = str(source.get("runtime_data") or "z.automazioni/data/runtime/runtime_exports.json").strip()
    payload = {
        "generated_by": "render_runtime_exports",
        "generated_from": SOURCE,
        "source": SOURCE,
        "purpose": "Registry runtime dei moduli caricati dal bridge DataviewJS e delle dipendenze CommonJS Templater.",
        "commonjs_runtime": commonjs_runtime(source),
        "runtime_modules": runtime_modules(source),
    }

    out_path = ROOT / target
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(f"Runtime exports data renderizzato: {target}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
